hera_cal/apply_cal.py

```diff
@@ -71,84 +78,103 @@ def recalibrate_in_place(data, data_flags, new_gains, cal_flags, old_gains=None,
-    new_gains, new_flags = io.load_cal(uvc)
+    hc = io.HERACal(new_calibration)
+    new_gains, new_flags, _, _ = hc.read()
+    add_to_history += ' NEW_CALFITS_HISTORY: ' + hc.history + '\n'
 
     # load old calibration solution
     if old_calibration is not None:
-        if isinstance(old_calibration, UVCal):
-            old_uvc = old_calibration
-        else:
-            old_uvc = UVCal()
-            old_uvc.read_calfits(old_calibration)
-        add_to_history += ' OLD_CALFITS_HISTORY: ' + old_uvc.history + '\n'
-        old_calibration, _ = io.load_cal(old_uvc)
+        old_hc = io.HERACal(old_calibration)
+        old_calibration, _, _, _ = old_hc.read()
+        add_to_history += ' OLD_CALFITS_HISTORY: ' + old_hc.history + '\n'
 
-    recalibrate_in_place(data, data_flags, new_gains, new_flags, old_gains=old_calibration, gain_convention=gain_convention)
-    io.update_vis(data_infilename, data_outfilename, filetype_in=filetype, filetype_out=filetype, data=data,
-                  flags=data_flags, add_to_history=add_to_history, clobber=clobber, **kwargs)
+    # partial loading and writing using uvh5
+    if nbl_per_load is not None:
+        if not (filetype_in is 'uvh5') or not (filetype_out is 'uvh5'):
+            raise NotImplementedError('Partial writing is not implemented for non-uvh5 I/O.')
+        hd = io.HERAData(data_infilename, filetype='uvh5')
+        for attribute, value in kwargs.items():
+            hd.__setattr__(attribute, value)
+        for data, data_flags, _ in hd.iterate_over_bls(Nbls=nbl_per_load):
+            for bl in data_flags.keys():
+                # apply band edge flags
+                data_flags[bl][:, 0:flag_nchan_low] = True
+                data_flags[bl][:, data_flags[bl].shape[1] - flag_nchan_high:] = True
+                # apply npz flags
+                if flags_npz is not None:
+                    data_flags[bl] = np.logical_or(data_flags[bl], npz_flag_dc[bl])
+            recalibrate_in_place(data, data_flags, new_gains, new_flags,
+                                 old_gains=old_calibration, gain_convention=gain_convention)
+            hd.partial_write(data_outfilename, data=data, flags=data_flags,
+                             inplace=True, clobber=clobber, add_to_history=add_to_history)
+
+    # full data loading and writing
+    else:
+        hd = io.HERAData(data_infilename, filetype=filetype_in)
+        hd.read()
+        # apply npz flags
+        if flags_npz is not None:
+            hd.flag_array = np.logical_or(npz_flags['flag_array'], hd.flag_array)
+        data, data_flags, _ = hd.build_datacontainers()
+        for bl in data_flags.keys():
+            # apply band edge flags
+            data_flags[bl][:, 0:flag_nchan_low] = True
+            data_flags[bl][:, data_flags[bl].shape[1] - flag_nchan_high:] = True
+            # apply npz flags
+            if flags_npz is not None:
+                data_flags[bl] = np.logical_or(data_flags[bl], npz_flag_dc[bl])
```

Comment:
Same question about keys in `npz_flag_dc`

Reply:
Same response as above.